<HTML><HEAD> <TITLE>MGI Data and Statistical Reports</TITLE> </HEAD><BODY
BGCOLOR="#DDDDDD">

<TABLE BORDER=0 CELLSPACING=0 CELLPADDING=0 BGCOLOR="#FFFFFF" WIDTH="100%">
<TR><TD><BR><H2>&nbsp;MGI Data and Statistical Reports</H2></TD> <TD
ALIGN="RIGHT"><A
HREF="http://www.informatics.jax.org/mgihome/other/copyright.shtml">Warranty
Disclaimer and Copyright Notice</A>&nbsp;&nbsp;&nbsp;</TD></TR> </TABLE><BR>

<TABLE BORDER BGCOLOR="#FFFFFF">

<TR><TD> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"> 
<TR><TH>Table of Contents</TH></TR> 
<TR><TD><A HREF="#ftp">FTP Server</A></TR></TD> 
<TR><TD><A HREF="#marker">Lists of all mouse genetic markers</A></TR></TD> 
<TR><TD><A HREF="#seq">Sequence Data</A></TR></TD> 
<TR><TD><A HREF="#orthology">Mammalian Orthology</A></TR></TD> 
<TR><TD><A HREF="#go">Gene Ontology Data</A></TR></TD>
<TR><TD><A HREF="#strain">Strains and Polymorphisms</A></TR></TD> 
<TR><TD><A HREF="#expression">Gene Expression</A></TR></TD> 
<TR><TD><A HREF="#pheno">Phenotypic Data</A></TR></TD> 
<TR><TD><A HREF="#statistics">General Statistics</A></TR></TD> 
<TR><TD><A HREF="#nomen">Nomenclature</A></TR></TD> 
<TR><TD><A HREF="#ref">References</A></TR></TD> 
<TR><TD><A HREF="#clone">Clones</A></TR></TD> </TABLE> </TD>

<TD VALIGN="CENTER"> &nbsp;&nbsp;&nbsp;Each entry contains the following information:<BR><BR>

<TABLE> 
<TR><TD><B>Summary description of data</B></TD></TR> 
<TR><TD> <TABLE BORDER CELLPADDING=2>
<TR><TD>Description of data in Field 1</TD>
<TD>Description of data in Field 2</TD>
<TD>etc.</TD></TR></TABLE> </TD></TR>
<TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<U>
<FONT COLOR="0000FF">filename on FTP server (link)</FONT></U> (additional details)</TD></TR> </TD></TR></TABLE>
</TD></TR></TABLE>

<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TD> <A
NAME="ftp"><B>FTP Server</B></A> <P>All reports described on this page are
within subdirectories of <B>/pub</B> on MGI's <A
HREF="ftp://ftp.informatics.jax.org">FTP Server</A>.<br><br> </TD></TR></TABLE>

<!-- -->

<A NAME="marker"></A><BR> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Lists of all mouse genetic markers</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. Sorted alphabetically by
marker symbol</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Marker Accession ID</TD><td
valign="top">Chromosome</TD><TD valign="top">cM Position</TD><td
valign="top">Marker Symbol</TD><TD valign="top">Status</TD><td
valign="top">Marker Name</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_List1.sql.rpt">MRK_List1.sql.rpt</A> (including withdrawn
symbols)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_List2.sql.rpt">MRK_List2.sql.rpt</A> (excluding withdrawn
symbols)</A></TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. Sorted numerically by
chromosome</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Marker Accession ID</TD><td
valign="top">Chromosome</TD><TD valign="top">cM Position</TD><td
valign="top">Marker Symbol</TD><TD valign="top">Status</TD><td
valign="top">Marker Name</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_List3.sql.rpt">MRK_List3.sql.rpt</A> (including withdrawn
symbols)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_List4.sql.rpt">MRK_List4.sql.rpt</A> (excluding withdrawn
symbols)</TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>3. Including
synonyms</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td
valign="top">MGI Marker Accession ID</TD><TD valign="top">Chromosome</TD><td
valign="top">cM Position</TD><TD valign="top">Marker Symbol</TD><td
valign="top">Synonym</TD></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_Synonym.sql.rpt">MRK_Synonym.sql.rpt</A></TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>4. Sorted by Marker Symbol
(tab delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Marker Accession ID</TD><td
valign="top">Marker Symbol</TD></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_Dump1.rpt">MRK_Dump1.rpt</A> (excluding withdrawn symbols)</TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>5. Sorted by Marker Symbol
(tab delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Marker Accession ID</TD><td
valign="top">Marker Symbol</TD><TD valign="top">Marker Name</TD><td
valign="top">cM Position</TD><TD valign="top">Chromosome</TD><td
valign="top">Marker Type</TD></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_Dump2.rpt">MRK_Dump2.rpt</A> (excluding withdrawn symbols)</TD></TR>
</TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="seq"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Sequence Data</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. MGI marker associations
to GenBank sequence information (tab-delimited)</B></TD></TR> <TR><TD
valign="top"> <TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker
ID</TD><TD valign="top">Marker Symbol</TD><td valign="top">Status</TD><td
valign="top">Marker Type</TD><TD valign="top">Marker Name</TD><td
valign="top">cM Position</TD><TD valign="top">Chromosome</TD> <td
valign="top">GenBank Accession IDs<BR>(space-separated)</TD><TD
valign="top">Unigene ID<BR>(if any)</TD><TD valign="top">RefSeq ID<BR>(if
any)</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_Sequence.rpt">MRK_Sequence.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. MGI marker associations
to SWISS-PROT protein IDs (tab-delimited)</B></TD></TR> <TR><TD valign="top">
<TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td
valign="top">Marker Symbol</TD><TD valign="top">Status</TD><td
valign="top">Marker Name</TD><TD valign="top">cM Position</TD><td
valign="top">Chromosome</TD> <TD valign="top">SWISS-PROT Protein<BR>Accession
IDs<BR>(space-separated)</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_SwissProt.rpt">MRK_SwissProt.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>3. MGI marker associations
to SWISS-PROT and TrEMBL protein IDs (tab-delimited)</B></TD></TR> <TR><TD
valign="top"> <TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker
ID</TD><TD valign="top">Marker Symbol</TD><td valign="top">Status</TD><td
valign="top">Marker Name</TD><TD valign="top">cM Position</TD><td
valign="top">Chromosome</TD> <TD valign="top">SWISS-PROT/TrEMBL
Protein<BR>Accession IDs<BR>(space-separated)</TD></TR></TABLE> </TD></TR>
<TR><TD valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_SwissProt_TrEMBL.rpt">MRK_SwissProt_TrEMBL.rpt</A></TD></TR>
</TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE><B>4. MGI marker associations to LocusLink (tab-
delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td valign="top">Marker
Symbol</TD><TD valign="top">Status</TD><td valign="top">Marker Name</TD><td
valign="top">cM Position</TD><TD valign="top">Chromosome</TD> <td
valign="top">Type<BR><FONT SIZE=-1>Gene<BR>DNA Segment<BR>Other Genome
Feature<BR>Complex/Cluster/Region<BR>microRNA</FONT></TD> <TD
valign="top">Secondary<BR>Accession IDs<BR>(|-delimited)</TD><TD
valign="top">LocusLink ID</TD><TD valign="top">Synonyms<BR>(|-
delimited)</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_LocusLink.rpt">MRK_LocusLink.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>5. MGI marker associations
to ENSEMBL sequence information (tab-delimited)</B></TD></TR> <TR><TD
valign="top"> <TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker
ID</TD><TD valign="top">Marker Symbol</TD><td valign="top">Marker Name</TD><td
valign="top">cM Position</TD><TD valign="top">Chromosome</TD> <td
valign="top">ENSEMBL Accession ID</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_ENSEMBL.rpt">MRK_ENSEMBL.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE><B>6. Primer Sequences</B></TD></TR> <TR><td
valign="top"> <TABLE BORDER CELLPADDING=2><TR><TD valign="top">Marker
Symbol</TD><TD valign="top">Marker Name</TD><td valign="top">Primer Name</TD><td
valign="top">MGI Marker Accession ID</TD><TD valign="top">MGI Primer Pair
ID</TD><TD valign="top">Primer 1 Sequence</TD><td valign="top">Primer 2
Sequence</TD><TD valign="top">Amplimer Size</TD><td
valign="top">Chromosome</TD><TD valign="top">Map Position</TD></TR></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_PrimerSeq.rpt">PRB_PrimerSeq.rpt</A> </TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE><B>7. Mapped ESTs (tab-delimited)</B></TD></TR>
<TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td valign="top">EST MGI
Acc ID</TD><TD valign="top">EST GenBank ID</TD><td
valign="top">Chromosome</TD><TD valign="top">cM Position</TD><td
valign="top">Marker Symbol</TD><TD valign="top">Marker Name</TD></TR></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_ESTMapped.rpt">PRB_ESTMapped.rpt</A> </TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>8. MGI marker associations
to InterPro IDs (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE
BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td
valign="top">Marker Symbol</TD><TD valign="top">InterPro Accession IDs (space-
separated)</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_InterPro.rpt">MRK_InterPro.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>9. InterPro
Domains</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td
valign="top">InterPro ID</TD><TD valign="top">Domain Name</TD></TR></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MGI_InterProDomains.sql.rpt">MGI_InterProDomains.sql.rpt</A></TD></TR>
</TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>10. MGI marker associations
to RIKEN Clones (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE
BORDER CELLPADDING=2><TR><TD valign="top">RIKEN Clone ID</TD><td
valign="top">MGI Clone ID</TD><TD valign="top">GenBank Sequence ID</TD><td
valign="top">Problem Sequence?</TD><TD valign="top">MGI Marker ID</TD><td
valign="top">Marker Symbol</TD><TD valign="top">Marker Name</TD></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MGI_RIKEN.rpt">MGI_RIKEN.rpt</A></TD></TR> </TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE><B>11. MGI-DoTS and MGI-TIGR mouse gene index
associations (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td valign="top">DT/TC
ID</TD><TD valign="top">GenBank Sequence IDs (comma-separated)</TD></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_DT_via_GB_one_one">MGI_DT_via_GB_one_one</A> (one MGI marker
to one DoTs sequence)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_DT_via_GB_one_many">MGI_DT_via_GB_one_many</A> (one MGI
marker to many DoTs sequences)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_DT_via_GB_many_one">MGI_DT_via_GB_many_one</A> (many MGI
markers to one DoTs sequence)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_DT_via_GB_many_many">MGI_DT_via_GB_many_many</A> (many MGI
markers to many DoTs sequences)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_TC_via_GB_one_one">MGI_TC_via_GB_one_one</A> (one MGI marker
to one TIGR gene index)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_TC_via_GB_one_many">MGI_TC_via_GB_one_many</A> (one MGI
marker to many TIGR gene indexes)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_TC_via_GB_many_one">MGI_TC_via_GB_many_one</A> (many MGI
markers to one TIGR gene index)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="dotstigr/MGI_TC_via_GB_many_many">MGI_TC_via_GB_many_many</A> (many MGI
markers to many TIGR gene indexes)</TD></TR> </TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE><B>12. <A
HREF="../sequence_dbs/seq_dbs.current/index.html">MouseBLAST Sequence
Databases</A></B> (a link to the MouseBLAST Sequence Databases page)</TD></TR>
</TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="orthology"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Mammalian Orthology</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. Human and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Human Chromosome</TD><td valign="top">Human
LocusLink ID</TD><TD valign="top">Human Marker Symbol</TD><td valign="top">Mouse
MGI Accession ID</TD><TD valign="top">Mouse Chromosome</TD><td
valign="top">Mouse cM Position</TD> <TD valign="top">Mouse LocusLink ID</TD><td
valign="top">Mouse Marker Symbol</TD><TD valign="top">Mouse Marker
Name</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Human1.rpt">HMD_Human1.rpt</A> (sorted by human chromosome)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. Human and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Mouse MGI Accession ID</TD><td
valign="top">Mouse Chromosome</TD><TD valign="top">Mouse cM Position</TD><td
valign="top">Mouse LocusLink ID</TD><TD valign="top">Mouse Marker Symbol</TD><td
valign="top">Human Chromosome</TD> <TD valign="top">Human LocusLink ID</TD><td
valign="top">Human Marker Symbol</TD><TD valign="top">Mouse Marker
Name</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Human2.rpt">HMD_Human2.rpt</A> (sorted by mouse chromosome)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>3. Human and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Human LocusLink ID</TD><td valign="top">Human
Marker Symbol </TD><TD valign="top">Human Chromosome</TD><td valign="top">Mouse
MGI Accession ID</TD><TD valign="top">Mouse LocusLink ID</TD><td
valign="top">Mouse Marker Symbol</TD><TD valign="top">Mouse Chromosome</TD> <td
valign="top">Mouse cM Position</TD><TD valign="top">Mouse Cytogenetic
Band</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Human3.rpt">HMD_Human3.rpt</A> (sorted by human symbol)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>4. Human and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Mouse MGI Accession ID </TD><td
valign="top">Mouse LocusLink ID</TD><TD valign="top">Mouse Marker Symbol</TD><td
valign="top">Mouse Chromosome</TD><TD valign="top">Mouse cM Position</TD><td
valign="top">Mouse Cytogenetic Band</TD> <TD valign="top">Human LocusLink
ID</TD><TD valign="top">Human Marker Symbol</TD><td valign="top">Human
Chromosome</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Human4.rpt">HMD_Human4.rpt</A> (sorted by mouse symbol)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>5. Human and Mouse
Orthology (GDB) (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE
BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td
valign="top">Mouse Marker Symbol</TD><TD valign="top">Mouse Marker Name</TD><td
valign="top">GDB ID</TD><TD valign="top">GDB Human Symbol</TD><td
valign="top">LocusLink ID</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="GDB_Accession.rpt">GDB_Accession.rpt</A></TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE><B>6. Human and Mouse Orthology with Sequence
information (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR> <TD valign="top">Mouse Marker Symbol</TD> <td
valign="top">MGI Marker ID</TD> <TD valign="top">Mouse LocusLink ID</TD> <td
valign="top">Human Marker Symbol</TD> <TD valign="top">Human LocusLink ID</TD>
<TD valign="top">Mouse RefSeq ID</TD> <td valign="top">Human RefSeq ID</TD> <td
valign="top">Mouse SWISS-PROT IDs<BR>(if any, comma-separated)</TD> <TD
valign="top">Human SWISS-PROT ID<BR>(if any)</TD> <TD valign="top">Mouse GenBank
Accession IDs<BR>(if no RefSeq or SWISS-PROT, comma-separated)</TD> <TD
valign="top">Evidences used to support the mouse-human orthology<BR>(comma-
separated)</TD> </TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_HumanSequence.rpt">HMD_HumanSequence.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>7. Rat and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Rat Chromosome</TD><td valign="top">Rat
LocusLink ID</TD><TD valign="top">Rat marker Symbol</TD><td valign="top">Mouse
MGI Accession ID</TD><TD valign="top">Mouse Chromosome</TD><td
valign="top">Mouse cM Position</TD> <TD valign="top">Mouse LocusLink ID</TD><td
valign="top">Mouse marker Symbol</TD><TD valign="top">Mouse marker
Name</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Rat1.rpt">HMD_Rat1.rpt</A> (sorted by rat chromosome)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>8. Rat and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Mouse MGI Accession ID</TD><td
valign="top">Mouse Chromosome</TD><TD valign="top">Mouse cM Position</TD><td
valign="top">Mouse LocusLink ID</TD><TD valign="top">Mouse Marker Symbol</TD><td
valign="top">Rat Chromosome</TD> <TD valign="top">Rat LocusLink ID</TD><td
valign="top">Rat Marker Symbol</TD><TD valign="top">Mouse Marker
Name</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Rat2.rpt">HMD_Rat2.rpt</A> (sorted by mouse chromosome)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>9. Rat and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Rat LocusLink ID</TD><td valign="top">Rat
Marker Symbol </TD><TD valign="top">Rat Chromosome</TD><td valign="top">Mouse
MGI Accession ID</TD><TD valign="top">Mouse LocusLink ID</TD><td
valign="top">Mouse Marker Symbol</TD><TD valign="top">Mouse Chromosome</TD> <td
valign="top">Mouse cM Position</TD><TD valign="top">Mouse Cytogenetic
Band</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Rat3.rpt">HMD_Rat3.rpt</A> (sorted by rat symbol)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>10. Rat and Mouse
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Mouse MGI Accession ID </TD><td
valign="top">Mouse LocusLink ID</TD><TD valign="top">Mouse Marker Symbol</TD><td
valign="top">Mouse Chromosome</TD><TD valign="top">Mouse cM Position</TD><td
valign="top">Mouse Cytogenetic Band</TD> <TD valign="top">Rat LocusLink
ID</TD><TD valign="top">Rat Marker Symbol</TD><td valign="top">Rat
Chromosome</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_Rat4.rpt">HMD_Rat4.rpt</A> (sorted by mouse symbol)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>11. Rat and Human
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Rat Chromosome</TD><td valign="top">Rat
LocusLink ID</TD><TD valign="top">Rat Marker Symbol</TD><td valign="top">Human
Chromosome</TD> <TD valign="top">Human LocusLink ID</TD><td valign="top">Human
Marker Symbol</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_RatHuman1.rpt">HMD_RatHuman1.rpt</A> (sorted by rat
chromosome)</TD></TR> </TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>12. Rat and Human
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Human Chromosome</TD><td valign="top">Human
LocusLink ID</TD><TD valign="top">Human Marker Symbol</TD><td valign="top">Rat
Chromosome</TD> <TD valign="top">Rat LocusLink ID</TD><td valign="top">Rat
Marker Symbol</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_RatHuman2.rpt">HMD_RatHuman2.rpt</A> (sorted by human
chromosome)</TD></TR> </TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>13. Rat and Human
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Rat LocusLink ID</TD><td valign="top">Rat
Marker Symbol </TD><TD valign="top">Rat Chromosome</TD><td valign="top">Human
LocusLink ID</TD><TD valign="top">Human Marker Symbol</TD><td valign="top">Human
Chromosome</TD></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_RatHuman3.rpt">HMD_RatHuman3.rpt</A> (sorted by rat symbol)</TD></TR>
</TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>14. Rat and Human
Orthology</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Human LocusLink ID</TD><td valign="top">Human
Marker Symbol</TD><TD valign="top">Human Chromosome</TD> <td valign="top">Rat
LocusLink ID</TD><TD valign="top">Rat Marker Symbol</TD><td valign="top">Rat
Chromosome</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_RatHuman4.rpt">HMD_RatHuman4.rpt</A> (sorted by human
symbol)</TD></TR> </TABLE></TD></TR> 

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>15. Mouse/Human/Rat Oxford
Grid Counts (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Species 1</TD><td valign="top">Species
2</TD><TD valign="top">Species 1 Chromosome</TD><td valign="top">Species 2
Chromosome</TD><TD valign="top">Number of Orthologies</TD></TABLE> </TD></TR>
<TR><TD valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_OxfordGrid.rpt">HMD_OxfordGrid.rpt</A></TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>16. Mouse/Human Orthology
with Phenotype Annotations (tab-delimited)</B></TD></TR> <TR><TD valign="top">
<TABLE BORDER CELLPADDING=2><TR><TD valign="top">Human Marker Symbol</TD><td
valign="top">Human LocusLink ID</TD><TD valign="top">Mouse Marker Symbol</TD><td
valign="top">MGI Marker ID</TD><TD valign="top">Mammalian Phenotype
ID<BR>(space-separated)</TD></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_HumanPhenotype.rpt">HMD_HumanPhenotype.rpt</A></TD></TR>
</TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="go"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Gene Ontology Data</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. Gene Ontology (GO)
Annotations of Mouse Markers (tab-delimited) <BR>(<I>see also</I> <A
HREF="http://www.geneontology.org/doc/GO.annotation.html#file">GO Annotation
File format</A> at the Gene Ontology Consortium, and <A
HREF="go_terms.mgi">go_terms.mgi</A> to translate GO IDs to GO
terms)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td
valign="top">Database<BR>Designation</TD><TD valign="top">MGI Marker ID</TD><td
valign="top">Marker Symbol</TD><TD valign="top">NOT<BR>Designation</TD><td
valign="top">GO<BR>Term ID</TD><TD valign="top">MGI<BR>Reference ID</TD><td
valign="top">GO<BR>Evidence Code</TD><TD valign="top">Inferred From</TD> <td
valign="top">Ontology<BR><FONT SIZE=-1>P=Biological Process<BR>F=Molecular
Function<BR>C=Cellular Component</FONT></TD> <TD valign="top">Marker
Name</TD><TD valign="top">Marker Synonyms<BR>(if any)</TD><td
valign="top">Marker Type<BR><FONT SIZE=-
1>gene<BR>transcript<BR>protein</FONT></TD><TD valign="top">Taxon</TD><td
valign="top">Modification Date</TD><TD valign="top">Assigned
By</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="gene_association.mgi">gene_association.mgi</A></TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. References used in GO
annotations to mouse markers (tab-delimited)</B></TD></TR> <TR><TD valign="top">
<TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Reference ID</TD><td
valign="top">PubMed ID</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="go_refs.mgi">go_refs.mgi</A></TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>3. GO terms and GO IDs for
all three ontologies (Biological Process, Cellular Component, Molecular
Function) (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Ontology</TD><td valign="top">GO ID</TD><td
valign="top">GO Term</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="go_terms.mgi">go_terms.mgi</A> (only those terms annotated to MGI markers are included)</TD></TR> </TABLE></TD></TR>           

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>4. GO annotations to mouse
markers that have human orthologs (tab-delimited)</B></TD></TR> <TR><TD
valign="top"> <TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker
ID</TD><TD valign="top">Marker Symbol</TD><td valign="top">Marker Name</TD><td
valign="top">GO ID</TD><TD valign="top">GO Term</TD> <td valign="top">GO
Evidence Code</TD><TD valign="top">PubMed ID</TD><td valign="top">SWISS-
PROT</TD><TD valign="top">GenBank ID<BR>EMBL ID<BR>DDBJ ID</TD></TR></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_GOHuman.rpt">MRK_GOHuman.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>5. Marker/SWISS-PROT
Associations for Markers with GO Annotations (tab-delimited)</B></TD></TR>
<TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td valign="top">MGI
Marker ID</TD><TD valign="top">SWISS-PROT ID (;-delimited)</TD></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="gp2protein.mgi">gp2protein.mgi</A></TD></TR> </TD></TR> </TABLE></TD></TR>

</TABLE><BR>

<!-- --> <A name="strain"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Strains and Polymorphisms</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. Official Strain
Nomenclature (tab-delimited)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Strain ID</TD><td valign="top">Strain
Name</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MGI_Strain.rpt">MGI_Strain.rpt</A> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. List of Strains with
Polymorphism Data</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Strain Name</TD></TR></TABLE> </TD></TR>
<TR><TD valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_Strain1.sql.rpt">PRB_Strain1.sql.rpt</A>
<BR>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(sorted by
strains)</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_Strain2.sql.rpt">PRB_Strain2.sql.rpt</A>
<BR>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(sorted by number of
references associated with polymorphic information for a particular
strain)</TD></TR> </TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="expression"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Gene Expression</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. List of mouse tissues
(origin of cDNAs)</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">tissue</TD></TR></TABLE> </TD></TR> <TR><td
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_TissueMouse.sql.rpt">PRB_TissueMouse.sql.rpt</A></TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. MGI Genetic Markers with
GXD Literature Index Records (tab-delimited)</B></TD></TR> <TR><TD valign="top">
<TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td
valign="top">Marker Symbol</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_GXD.rpt">MRK_GXD.rpt</A> </TABLE></TD></TR>        

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>3. MGI Genetic Markers with GXD Assay Records (tab-delimited)</B></TD></TR> <TR><td valign="top"> <TABLE
BORDER CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td
valign="top">Marker Symbol</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_GXDAssay.rpt">MRK_GXDAssay.rpt</A> </TABLE></TD></TR>        

</TABLE><BR>

<!-- -->

<A name="pheno"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Phenotypic Data</TH></TR>

<TR><TD valign="top">Please note that the MGI Accession IDs of the Phenotype
Classification Terms have been replaced with the new Mammalian Phenotype
Ontology accession IDs (MP:####) as of January 28, 2003.  The MGI IDs are being
retained as secondary accession IDs for these terms.</TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. PhenoSlim
vocabulary</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MP Accession ID</TD><td
valign="top">PhenoSlim Term</TD><TD valign="top">Definition</TD><td
valign="top">Example</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="VOC_PhenoSlim.rpt">VOC_PhenoSlim.rpt</A> (tab-delimited, without column
headings)</TD></TR> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="VOC_PhenoSlim.sql.rpt">VOC_PhenoSlim.sql.rpt</A> (space-delimited; with
column headings)</TD></TR> </TD></TR> </TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. Mammalian Phenotype
Vocabulary in DAG-Edit Format</B></TD></TR> <TR><TD valign="top"> </TD></TR>
<TR><TD valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MPheno.ontology">MPheno.ontology</A> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MP.defs">MP.defs</A> </TABLE></TD></TR>        

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>3. Mouse/Human Orthology
with Phenotype Annotations (tab-delimited)</B></TD></TR> <TR><TD valign="top">
<TABLE BORDER CELLPADDING=2><TR><TD valign="top">Human Marker Symbol</TD><td
valign="top">Human LocusLink ID</TD><TD valign="top">Mouse Marker Symbol</TD><td
valign="top">MGI Marker ID</TD><TD valign="top">Mammalian Phenotype
ID<BR>(space-separated)</TD></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="HMD_HumanPhenotype.rpt">HMD_HumanPhenotype.rpt</A></TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>4. MGI marker associations to Phenotype Annotations and ENSEMBL sequence information (tab-delimited)</B></TD></TR> <TR><td valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">MGI Marker ID</TD><td valign="top">Marker Symbol</TD><td valign="top">Marker Name</TD><td valign="top">ENSEMBL Accession
ID</TD> <TD valign="top">Mammalian Phenotype ID</TD><td valign="top">Mammalian
Phenotype Term</TD><TD valign="top">PubMed ID of Marker/Phenotype Association
(comma-delimited)</TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_Ensembl_Pheno.rpt">MRK_Ensembl_Pheno.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="statistics"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>General Statistics</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. MGI General Statistics -
Number of:</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER CELLPADDING=2>
<TR><TD valign="top">References</TD></TR> <TR><td valign="top">Genetic
Markers</TD></TR> <TR><TD valign="top">Genes</TD></TR> <TR><td
valign="top">Genes with DNA Sequence Information</TD></TR> <TR><TD
valign="top">Genes with Protein Sequence Information</TD></TR> <TR><TD
valign="top">Genetic Markers Mapped</TD></TR> <TR><TD valign="top">Genes
Mapped</TD></TR> <TR><TD valign="top">Mouse/Human Orthologies</TD></TR> <TR><td
valign="top">Genes with Molecular Probes and Segments Data</TD></TR> <TR><TD
valign="top">Markers with Molecular Polymorphisms</TD></TR> <TR><TD
valign="top">Genes with Molecular Polymorphisms</TD></TR></TABLE> </TD></TR>
<TR><TD valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MGD_Stats.sql.rpt">MGD_Stats.sql.rpt</A></TD></TR> </TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="nomen"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Nomenclature</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. Nomenclature Events in
the Past Week</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Chromosome</TD><td valign="top">Marker
Symbol</TD><TD valign="top">Marker Name</TD><td valign="top">J Number</TD><td
valign="top">First Author</TD><TD valign="top">MGI<BR>Marker ID</TD><td
valign="top">GenBank<BR>Sequence ID</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="Nomenclature-current.html">Nomenclature-current.html</A></TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. Archived Nomenclature
Reports (individual reports as above)</B></TD></TR> <TR><TD valign="top"> <TABLE
BORDER CELLPADDING=2><TR><TD valign="top">Chromosome</TD><td valign="top">Marker
Symbol</TD><TD valign="top">Marker Name</TD><td valign="top">J Number</TD><td
valign="top">First Author</TD><TD valign="top">MGI<BR>Marker ID</TD><td
valign="top">GenBank<BR>Sequence ID</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="archive/nomen">archive/nomen</A></TD></TR> </TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="ref"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>References</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. Association of MGI Markers and PubMed IDs (tab-delimited)</B></TD></TR> <TR><td valign="top">
<TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI ID</TD><td
valign="top">Marker Symbol</TD><TD valign="top">Marker Name</TD><td
valign="top">Synonyms<BR>(|-delimited)</TD><TD valign="top">PubMed IDs<BR>(|-
delimited)</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="MRK_Reference.rpt">MRK_Reference.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. Association of MGI
Reference IDs and PubMed IDs (tab-delimited)</B></TD></TR> <TR><TD valign="top">
<TABLE BORDER CELLPADDING=2><TR><TD valign="top">MGI Reference ID</TD><td
valign="top">PubMed ID</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="BIB_PubMed.rpt">BIB_PubMed.rpt</A></TD></TR> </TD></TR> </TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A name="clone"></A> <TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF"
WIDTH="100%"><TR><TH>Clones</TH></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>1. MGI Clone
Libraries</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER
CELLPADDING=2><TR><TD valign="top">Library Name</TD></TR></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_CloneLibrary.rpt">PRB_CloneLibrary.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

<TR><TD valign="top"><TABLE> <TR><td valign="top"><B>2. MGI Clone
Sets</B></TD></TR> <TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td
valign="top">Segment ID</TD><TD valign="top">Segment Name</TD><td
valign="top">Marker ID</TD><TD valign="top">Symbol</TD></TR></TABLE>
</TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_CloneSetImage.rpt">PRB_CloneSetImage.rpt</A></TD></TR> </TD></TR>
<TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td valign="top">Segment
ID</TD><TD valign="top">Segment Name</TD><td valign="top">Marker
ID</TD><TD valign="top">Marker Symbol</TD><td
valign="top">Database</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_CloneSetNIA.rpt">PRB_CloneSetNIA.rpt</A></TD></TR> </TD></TR> <TR><TD
valign="top"> <TABLE BORDER CELLPADDING=2><TR><TD valign="top">Segment
ID</TD><TD valign="top">Segment Name</TD><td valign="top">Marker
ID</TD><TD valign="top">Marker Symbol</TD></TR></TABLE> </TD></TR> <TR><td
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_CloneSetRPCI23.rpt">PRB_CloneSetRPCI23.rpt</A></TD></TR> </TD></TR>
<TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td valign="top">Segment
ID</TD><TD valign="top">Segment Name</TD><td valign="top">Marker
ID</TD><TD valign="top">Marker Symbol</TD></TR></TABLE> </TD></TR> <TR><td
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_CloneSetRPCI24.rpt">PRB_CloneSetRPCI24.rpt</A></TD></TR> </TD></TR>
<TR><TD valign="top"> <TABLE BORDER CELLPADDING=2><TR><td valign="top">Segment
ID</TD><TD valign="top">Segment Name</TD><td valign="top">Marker
ID</TD><TD valign="top">Marker Symbol</TD><td
valign="top">Database</TD></TR></TABLE> </TD></TR> <TR><TD
valign="top">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A
HREF="PRB_CloneSetRiken.rpt">PRB_CloneSetRiken.rpt</A></TD></TR> </TD></TR>
</TABLE></TD></TR>

</TABLE><BR>

<!-- -->


<PRE>










</PRE>

</BODY></HTML>
