<HTML><HEAD>
<TITLE>MGI Data and Statistical Reports</TITLE>
</HEAD><BODY BGCOLOR="#DDDDDD">

<TABLE BORDER=0 CELLSPACING=0 CELLPADDING=0 BGCOLOR="#FFFFFF" WIDTH="100%">
<TR><TD><BR><H2>&nbsp;MGI Data and Statistical Reports</H2></TD>
<TD ALIGN="RIGHT"><A HREF="http://www.informatics.jax.org/mgihome/other/copyright.shtml">Warranty Disclaimer and Copyright Notice</A>&nbsp;&nbsp;&nbsp;</TD></TR>
</TABLE><BR>

<TABLE BORDER BGCOLOR="#FFFFFF">

<TR><TD>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF">
<TR><TH>Table of Contents</TH></TR>
<TR><TD><A HREF="#ftp">FTP Server</A></TR></TD>
<TR><TD><A HREF="#marker">Lists of all mouse genetic markers</A></TR></TD>
<TR><TD><A HREF="#seq">Sequence Data</A></TR></TD>
<TR><TD><A HREF="#homology">Mammalian Orthology</A></TR></TD>
<TR><TD><A HREF="#go">Gene Ontology Data</A></TR></TD>
<TR><TD><A HREF="#strain">Strains and Polymorphisms</A></TR></TD>
<TR><TD><A HREF="#expression">Gene Expression</A></TR></TD>
<TR><TD><A HREF="#pheno">Phenotypic Data</A></TR></TD>
<TR><TD><A HREF="#statistics">General Statistics</A></TR></TD>
<TR><TD><A HREF="#nomen">Nomenclature</A></TR></TD>
<TR><TD><A HREF="#ref">References</A></TR></TD>
</TABLE>
</TD>

<TD VALIGN="CENTER">
&nbsp;&nbsp;&nbsp;Each entry contains information as follows:<BR><BR>

<TABLE>
        <TR><TD><B>Summary description of data</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Description of data in Field 1</TD><TD>Description of data in Field 2</TD><TD>etc.</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<U><FONT COLOR="0000FF">filename on
 FTP server (link)</FONT></U> (additional details)</TD></TR>
</TD></TR></TABLE>

</TD></TR></TABLE>

<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TD>
<A NAME="ftp"><B>FTP Server</B></A><BR><BR>
<P>All reports described on this page are within subdirectories of <B>/pub</B> on MGI's <A HREF="ftp://ftp.informatics.jax.org">FTP Server</A>.<BR><BR>
</TD></TR></TABLE>

<!-- -->

<A NAME="marker"></A><BR>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Lists of all mouse genetic markers</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. Sorted alphabetically by symbol</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Accession ID</TD><TD>Chromosome</TD><TD>cM Position</TD><TD>Symbol</TD><TD>Name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_List1.sql.rpt">MRK_List1.sql.rpt</A> (including withdrawn symbols)</TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_List2.sql.rpt">MRK_List2.sql.rpt</A> (excluding withdrawn symbols)</A></TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE>
        <TR><TD><B>2. Sorted numerically by chromosome</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Accession ID</TD><TD>Chromosome</TD><TD>cM Position</TD><TD>Symbol</TD><TD>Name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_List3.sql.rpt">MRK_List3.sql.rpt</A> (including withdrawn symbols)</TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_List4.sql.rpt">MRK_List4.sql.rpt</A> (excluding withdrawn symbols)</TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE>
        <TR><TD><B>3. Including synonyms</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Accession ID</TD><TD>Chromosome</TD><TD>cM Position</TD><TD>Symbol</TD><TD>Synonym</TD></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_Synonym.sql.rpt">MRK_Synonym.sql.rpt</A></TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE>
        <TR><TD><B>4. Sorted by MGI Accession ID (tab delimited)</B></TD></TR> 
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Accession ID</TD><TD>Symbol</TD></TABLE>
        </TD></TR>         
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_Dump1.rpt">MRK_Dump1.rpt</A> (excluding withdrawn symbols)</TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE>
        <TR><TD><B>5. Sorted by Symbol (tab delimited)</B></TD></TR>    
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Accession ID</TD><TD>Symbol</TD><TD>Name</TD><TD>cM Position</TD><TD>Chromosome</TD><TD>Marker Type</TD></TABLE>
        </TD></TR>   
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_Dump2.rpt">MRK_Dump2.rpt</A> (excluding withdrawn symbols)</TD></TR>
    </TABLE></TD></TR>
    
</TABLE><BR>

<!-- -->

<A NAME="seq"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Sequence Data</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. MGI markers associations to GenBank sequence information (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Marker ID</TD><TD>Symbol</TD><TD>Name</TD><TD>cM position</TD><TD>Chromosome</TD>
        <TD>GenBank Accession IDs<BR>(space-separated)</TD><TD>Unigene ID<BR>(if any)</TD><TD>RefSeq ID<BR>(if any)</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_Sequence.rpt">MRK_Sequence.rpt</A></TD></TR>
        </TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE>
        <TR><TD><B>2. MGI marker associations to SwissProt protein IDs (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Marker ID</TD><TD>Symbol</TD><TD>Name</TD><TD>cM position</TD><TD>Chromosome</TD>
        <TD>SwissProt Protein<BR>Accession IDs<BR>(space-separated)</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_SwissProt.rpt">MRK_SwissProt.rpt</A></TD></TR>
        </TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE><B>3. MGI marker associations to LocusLink (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Marker ID</TD><TD>Symbol</TD><TD>Name</TD><TD>cM position</TD><TD>Chromosome</TD>
        <TD>Type<BR><FONT SIZE=-1>Gene<BR>DNA Segment<BR>Other Genome Feature<BR>Complex/Cluster/Region</FONT></TD>
        <TD>Secondary<BR>Accession IDs<BR>(|-delimited)</TD><TD>LocusLink ID</TD><TD>Synonyms<BR>(|-delimited)</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_LocusLink.rpt">MRK_LocusLink.rpt</A></TD></TR>
        </TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE><B>4. Primer Sequences</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Primer Name</TD><TD>Marker Symbol</TD><TD>Primer 1 Sequence</TD><TD>Primer 2 Sequence</TD><TD>Amplimer Size</TD><TD>MGI Accession ID</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="PRB_PrimerSeq.rpt">PRB_PrimerSeq.rpt</A>
        </TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE><B>5. Mapped ESTs (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>EST MGI Acc ID</TD><TD>EST GenBank ID</TD><TD>chromosome</TD><TD>cM position</TD><TD>marker symbol</TD><TD>marker name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="PRB_ESTMapped.rpt">PRB_ESTMapped.rpt</A>
        </TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE><B>6. MGI-DoTS and MGI-TIGR mouse gene index Associations (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Marker ID</TD><TD>DT/TC ID</TD><TD>GenBank Sequence IDs (comma-separated)</TD></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_DT_via_GB_one_one">MGI_DT_via_GB_one_one</A></TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_DT_via_GB_one_many">MGI_DT_via_GB_one_many</A></TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_DT_via_GB_many_one">MGI_DT_via_GB_many_one</A></TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_DT_via_GB_many_many">MGI_DT_via_GB_many_many</A></TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_TC_via_GB_one_one">MGI_TC_via_GB_one_one</A></TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_TC_via_GB_one_many">MGI_TC_via_GB_one_many</A></TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_TC_via_GB_many_one">MGI_TC_via_GB_many_one</A></TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="../datasets/dotstigr/MGI_TC_via_GB_many_many">MGI_TC_via_GB_many_many</A></TD></TR>
        </TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE><B>7. <A HREF="../sequence_dbs/seq_dbs.current/index.html">MouseBLAST Sequence Databases</A></B></TD></TR>
    </TABLE></TD></TR>
    
</TABLE><BR>

<!-- -->

<A NAME="homology"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Mammalian Orthology</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. Human and Mouse Orthology</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Human<BR>Chromosome</TD><TD>Human<BR>Symbol</TD><TD>Mouse MGI<BR>Accession ID</TD><TD>Mouse<BR>Chromosome</TD><TD>Mouse<BR>cM Position</TD>
        <TD>Mouse<BR>Symbol</TD><TD>Mouse<BR>Name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Human1.rpt">HMD_Human1.rpt</A> (sorted by human chromosome)</TD></TR>
    </TABLE></TD></TR> 
     
     <TR><TD><TABLE>
     <TR><TD><B>2. Human and Mouse Orthology</B></TD></TR>   
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Mouse MGI<BR>Accession ID</TD><TD>Mouse<BR>Chromosome</TD><TD>Mouse<BR>cM Position</TD><TD>Mouse<BR>Symbol</TD><TD>Human<BR>Chromosome</TD>
        <TD>Human<BR>Symbol</TD><TD>Mouse<BR>Name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Human2.rpt">HMD_Human2.rpt</A> (sorted by mouse chromosome)</TD></TR>
    </TABLE></TD></TR> 
    
    <TR><TD><TABLE>
    <TR><TD><B>3. Human and Mouse Orthology</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Human<BR>Symbol<BR></TD><TD>Human<BR>Chromosome</TD><TD>Mouse MGI<BR>Accession ID</TD><TD>Mouse<BR>Symbol</TD><TD>Mouse<BR>Chromosome</TD>
        <TD>Mouse<BR>cM Position</TD><TD>Mouse<BR>Band</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Human3.rpt">HMD_Human3.rpt</A> (sorted by human symbol)</TD></TR>
    </TABLE></TD></TR> 
    
    <TR><TD><TABLE>
    <TR><TD><B>4. Human and Mouse Orthology</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Mouse MGI<BR>Accession ID<BR></TD><TD>Mouse<BR>Symbol</TD><TD>Mouse<BR>Chromosome</TD><TD>Mouse<BR>cM Position</TD><TD>Mouse<BR>Band</TD>
        <TD>Human<BR>Symbol</TD><TD>Human<BR>Chromosome</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Human4.rpt">HMD_Human4.rpt</A> (sorted by mouse symbol)</TD></TR>
    </TABLE></TD></TR> 

    <TR><TD><TABLE>
        <TR><TD><B>5. Human and Mouse Orthology (GDB) (tab-delimited)</B></TD></TR>        
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI<BR>Marker ID</TD><TD>Mouse<BR>Symbol</TD><TD>Mouse<BR>Name</TD><TD>GDB ID</TD><TD>GDB<BR>Human Symbol</TD><TD>LocusLink ID</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="GDB_Accession.rpt">GDB_Accession.rpt</A></TD></TR>
    </TABLE></TD></TR>
        
    <TR><TD><TABLE><B>6. Human and Mouse Orthology with Sequence information (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR>
		<TD>Mouse Symbol</TD>
		<TD>MGI Marker ID</TD>
		<TD>Human Symbol</TD>
		<TD>Human LocusLink ID</TD>
		<TD>Mouse RefSeq ID</TD>
		<TD>Human RefSeq ID</TD>
		<TD>Mouse SwissProt ID</TD>
		<TD>Human SwissProt ID</TD>
		<TD>Mouse GenBank Accession ID (if no RefSeq or SwissProt)</TD>
		<TD>Evidence used to support the mouse-human orthology</TD>
	</TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_HumanSeq.rpt">HMD_HumanSeq.rpt</A></TD></TR>
        </TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE>
        <TR><TD><B>7. Rat and Mouse Orthology</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Rat<BR>Chromosome</TD><TD>Rat<BR>Symbol</TD><TD>Mouse<BR>Chromosome</TD><TD>Mouse<BR>cM Position</TD>
        <TD>Mouse<BR>Symbol</TD><TD>Mouse<BR>Name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Rat1.sql.rpt">HMD_Rat1.sql.rpt</A> (sorted by rat chromosome)</TD></TR>
    </TABLE></TD></TR> 

    <TR><TD><TABLE>
        <TR><TD><B>8. Rat and Mouse Orthology</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Mouse<BR>Chromosome</TD><TD>Mouse<BR>cM Position</TD><TD>Mouse<BR>Symbol</TD><TD>Rat<BR>Chromosome</TD>
        <TD>Rat<BR>Symbol</TD><TD>Mouse<BR>Name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Rat2.sql.rpt">HMD_Rat2.sql.rpt</A> (sorted by mouse chromosome)</TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE>
        <TR><TD><B>9. Rat and Mouse Orthology</B></TD></TR>        
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Rat<BR>Symbol<BR></TD><TD>Rat<BR>Chromosome</TD><TD>Mouse<BR>Symbol</TD><TD>Mouse<BR>Chromosome</TD>
        <TD>Mouse<BR>cM Position</TD><TD>Mouse<BR>Band</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Rat3.sql.rpt">HMD_Rat3.sql.rpt</A> (sorted by rat symbol)</TD></TR>
    </TABLE></TD></TR>
    
    <TR><TD><TABLE>
        <TR><TD><B>10. Rat and Mouse Orthology</B></TD></TR>        
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Mouse<BR>Symbol</TD><TD>Mouse<BR>Chromosome</TD><TD>Mouse<BR>cM Position</TD><TD>Mouse<BR>Band</TD>
        <TD>Rat<BR>Symbol</TD><TD>Rat<BR>Chromosome</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="HMD_Rat4.sql.rpt">HMD_Rat4.sql.rpt</A> (sorted by mouse symbol)</TD></TR>
    </TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A NAME="go"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Gene Ontology Data</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. Gene Ontology (GO) Annotations of Mouse Markers (tab-delimited) 
        <BR>(<I>see also</I> <A HREF="http://www.geneontology.org/doc/GO.annotation.html#file">GO Annotation File format</A> at the Gene Ontology Consortium, and 
        <A HREF="go_terms.mgi">go_terms.mgi</A> to translate GO IDs to GO terms)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI</TD><TD>MGI<BR>Marker ID</TD><TD>Gene Symbol</TD><TD>GO<BR>Term ID</TD><TD>MGI<BR>Reference ID</TD><TD>GO<BR>Evidence Code</TD>
        <TD>Ontology<BR><FONT SIZE=-1>P=Biological Process<BR>F=Molecular Function<BR>C=Cellular Component</FONT></TD>
        <TD>Name</TD><TD>Synonyms<BR>(if any)</TD><TD>Object type<BR><FONT SIZE=-1>gene<BR>transcript<BR>protein</FONT></TD><TD>Taxon</TD><TD>Date</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="gene_association.mgi">gene_association.mgi</A></TD></TR>
    </TABLE></TD></TR>
        
    <TR><TD><TABLE>
        <TR><TD><B>2. References used in GO annotations to mouse markers (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Reference ID</TD><TD>PubMed ID</TD></TR></TABLE>
        </TD></TR> 
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="go_refs.mgi">go_refs.mgi</A></TD></TR>
    </TABLE></TD></TR>
        
    <TR><TD><TABLE>
        <TR><TD><B>3. GO terms and GO IDs for all three ontologies (Biological Process, Cellular Component, Molecular Function)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Ontology</TD><TD>GO ID</TD><TD>GO Term</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="go_terms.mgi">go_terms.mgi</A> (only those terms annotated to MGI genes are shown)</TD></TR>
    </TABLE></TD></TR>           

    <TR><TD><TABLE>
        <TR><TD><B>4. GO annotations to mouse markers that have human orthologs</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Marker ID</TD><TD>Symbol</TD><TD>Name</TD><TD>GO ID</TD><TD>GO Term</TD>
        <TD>GO Evidence Code</TD><TD>PubMed ID</TD><TD>SWISS-PROT/<BR>TREMBL ID</TD><TD>GenBank ID<BR>EMBL ID<BR>DDBJ ID</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_GOHuman.rpt">MRK_GOHuman.rpt</A></TD></TR>
        </TD></TR>
    </TABLE></TD></TR>
                
</TABLE><BR>

<!-- -->
<A NAME="strain"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Strains and Polymorphisms</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. List of Strains with Polymorphism Data</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Strain Name</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="PRB_Strain1.sql.rpt">PRB_Strain1.sql.rpt</A>
        <BR>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(sorted by strains)</TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="PRB_Strain2.sql.rpt">PRB_Strain2.sql.rpt</A>
        <BR>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(sorted by number of references associated with polymorphic information for a particular strain)</TD></TR>
    </TABLE></TD></TR>
    
</TABLE><BR>

<!-- -->

<A NAME="expression"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Gene Expression</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. List of mouse tissues (origin of cDNAs)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>tissue</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="PRB_TissueMouse.sql.rpt">PRB_TissueMouse.sql.rpt</A></TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE>
        <TR><TD><B>2. Alternate Transcripts</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Mouse Symbol</TD><TD>Chromosome</TD><TD>J Number</TD><TD>Human Symbol</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="PRB_AltTranscripts.rpt">PRB_AltTranscripts.rpt</A></TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE>
        <TR><TD><B>4. MGI Genes with GXD Literature Index Records</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Gene ID</TD><TD>Gene Symbol</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_GXD.rpt">MRK_GXD.rpt</A>
    </TABLE></TD></TR>        
        
    <TR><TD><TABLE>
        <TR><TD><B>5. MGI Genes with GXD Assay Records</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Gene ID</TD><TD>Gene Symbol</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MRK_GXDAssay.rpt">MRK_GXDAssay.rpt</A>
    </TABLE></TD></TR>        
                
</TABLE><BR>

<!-- -->

<A NAME="pheno"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Phenotypic Data</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. PhenoSlim vocabulary</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Accession ID</TD><TD>PhenoSlim Term</TD><TD>Definition</TD><TD>Example</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="VOC_PhenoSlim.rpt">VOC_PhenoSlim.rpt</A> (tab-delimited, without column headings)</TD></TR>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="VOC_PhenoSlim.sql.rpt">VOC_PhenoSlim.sql.rpt</A> (with column headings)</TD></TR>
        </TD></TR>
    </TABLE></TD></TR>

</TABLE><BR>

<!-- -->

<A NAME="statistics"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>General Statistics</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. MGI General Statistics - Number of:</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2>
        <TR><TD>References</TD></TR>
        <TR><TD>Genetic Markers</TD></TR>
        <TR><TD>Genes</TD></TR>
        <TR><TD>Genes with DNA Sequence Information</TD></TR>
        <TR><TD>Genes with Protein Sequence Information</TD></TR>
        <TR><TD>Genetic Markers Mapped</TD></TR>
        <TR><TD>Genes Mapped</TD></TR>
        <TR><TD>Mouse/Human Orthologies</TD></TR>
        <TR><TD>Genes with Molecular Probes and Segments Data</TD></TR>
        <TR><TD>Markers with Molecular Polymorphisms</TD></TR>
        <TR><TD>Genes with Molecular Polymorphisms</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="MGD_Stats.sql.rpt">MGD_Stats.sql.rpt</A></TD></TR>
    </TABLE></TD></TR>
    
</TABLE><BR>

<!-- -->

<A NAME="nomen"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>Nomenclature</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. Nomenclature Events in the Past Week</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Chromosome</TD><TD>Symbol</TD><TD>Name</TD><TD>J Number</TD><TD>First Author</TD><TD>MGI<BR>Marker ID</TD><TD>GenBank<BR>Sequence ID</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="Nomenclature-current.html">Nomenclature-current.html</A></TD></TR>
    </TABLE></TD></TR>

    <TR><TD><TABLE>
        <TR><TD><B>2. Archived Nomenclature Reports (individual reports as above)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>Chromosome</TD><TD>Symbol</TD><TD>Name</TD><TD>J Number</TD><TD>First Author</TD><TD>MGI<BR>Marker ID</TD><TD>GenBank<BR>Sequence ID</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="archive/nomen">archive/nomen</A></TD></TR>
    </TABLE></TD></TR>
    
</TABLE><BR>

<!-- -->

<A NAME="ref"></A>
<TABLE BORDER CELLPADDING=2 BGCOLOR="#FFFFFF" WIDTH="100%"><TR><TH>References</TH></TR>

    <TR><TD><TABLE>
        <TR><TD><B>1. Association of MGI Reference IDs and PubMed IDs (tab-delimited)</B></TD></TR>
        <TR><TD>
        <TABLE BORDER CELLPADDING=2><TR><TD>MGI Reference ID</TD><TD>PubMed ID</TD></TR></TABLE>
        </TD></TR>
        <TR><TD>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="BIB_PubMed.rpt">BIB_PubMed.rpt</A></TD></TR>
        </TD></TR>
    </TABLE></TD></TR>
    
</TABLE><BR>

<!-- -->


<PRE>















































</PRE>

</BODY></HTML>
